server/contexts/ctxerr/ctxerr.go

```diff
@@ -13,103 +13,231 @@ package ctxerr
 
 import (
 	"context"
+	"encoding/json"
 	"errors"
 	"fmt"
 	"strings"
 	"time"
-
-	"github.com/fleetdm/fleet/v4/server/errorstore"
-	"github.com/rotisserie/eris" //nolint:depguard
 )
 
 type key int
 
 const errHandlerKey key = 0
 
-// Cause returns the root error in err's chain.
-func Cause(err error) error {
-	// Until we use only ctxerr.Wrap, we have a mix of pkg/errors, fmt.Errorf
-	// and eris.Wrap (via ctxerr.Wrap). pkg/errors.Cause looks for a Cause()
-	// method, while eris.Cause looks for the stdlib-compliant Unwrap(). So
-	// implement a custom Cause that checks for both until a root is found.
+// Defining here for testing purposes
+var nowFn = time.Now
+
+// FleetError is the error implementation used by this package.
+type FleetError struct {
+	msg   string          // error message to be prepended to cause
+	stack StackTracer     // stack trace where this error was created
+	cause error           // original error that caused this error if non-nil
+	data  json.RawMessage // additional metadata about the error (timestamps, etc)
+}
+
+type fleetErrorJSON struct {
+	Message string          `json:"message,omitempty"`
+	Data    json.RawMessage `json:"data,omitempty"`
+	Stack   []string        `json:"stack,omitempty"`
+}
+
+// Error implements the error interface.
+func (e *FleetError) Error() string {
+	if e.cause == nil {
+		return e.msg
+	}
+	return fmt.Sprintf("%s: %s", e.msg, e.cause.Error())
+}
+
+// Unwrap implements the error Unwrap interface introduced in go1.13.
+func (e *FleetError) Unwrap() error {
+	return e.cause
+}
+
+// MarshalJSON implements the marshaller interface, giving us control on how
+// errors are json-encoded
+func (e *FleetError) MarshalJSON() ([]byte, error) {
+	return json.Marshal(fleetErrorJSON{
+		Message: e.msg,
+		Data:    e.data,
+		Stack:   e.stack.List(),
+	})
+}
+
+// Stack returns a call stack for the error
+func (e *FleetError) Stack() []string {
+	return e.stack.List()
+}
+
+// setMetadata adds common metadata attributes to the `data` map provided.
+// NOTE: this will mutate the data provided and override other values with the same keys.
+func setMetadata(ctx context.Context, data map[string]interface{}) map[string]interface{} {
+	if data == nil {
+		data = map[string]interface{}{}
+	}
 
-	var cerr interface {
-		Cause() error
+	// TODO: add more metadata from ctx
+	data["timestamp"] = nowFn().Format(time.RFC3339)
+	return data
+}
+
+func encodeData(ctx context.Context, data map[string]interface{}, augment bool) json.RawMessage {
+	if augment {
+		data = setMetadata(ctx, data)
+	}
+
+	encoded, err := json.Marshal(data)
+	if err != nil {
+		msg := fmt.Sprintf(`{"error": "there was an error encoding additional data: %s"}`, err.Error())
+		encoded = json.RawMessage(msg)
+	}
```

Comment:
for reference, this is done for the reasons discussed here: https://github.com/fleetdm/fleet/pull/5733#discussion_r874750279.

I'm open to any other ideas about this, I really don't want to panic or return another error when wrapping/creating a new error.